File: 2021_18.py
import sys;
import logging;
import re;
import math;

AOC_DAY=14
AOC_YEAR=2021


pairherere = re.compile('^(\d+),(\d+)(.*)');
#rightmostdigitre = re.compile('.*(\d+)([^\d]*)');
#leftmostdigitre = re.compile('[^\d]*(\d+)(.*)');
rightmostdigitre = re.compile('.*?(\d+)([^\d]*)$');
leftmostdigitre = re.compile('^[^\d]*(\d+)(.*?)');

def explode(snailfish, logger):
    logger.debug(f'explode: {snailfish}');
    # find the leftmost depth-4 pair and explode it.
    # work with the text representation of the number (no spaces)
    did = False
    st = str(snailfish).replace(' ','');
    depth = 0;
    last_number_index = None;
    for index, item in enumerate(st):
        if '[' == item:
            depth += 1;
        elif ']' == item:
            depth -= 1;
        elif depth > 4:
            logger.debug(f'explode: depth {depth} at {index} "{st[index:]}"');
            # find out if there's a pair here
            match = pairherere.match(st[index:]);
            if(match):
                # explode this pair
                stfirsthalf = st[:index-1];
                stlasthalf = st[index + match.start(3)+1:];
                # new string will be firsthalf 0 lasthalf
                # add first in matched pair to rightmost digit preceding match
                rightmostdigit_match = rightmostdigitre.match(stfirsthalf);
                if(rightmostdigit_match):
                    stfirsthalf = stfirsthalf[:rightmostdigit_match.start(1)] + \
                            str(int(match.group(1)) + int(rightmostdigit_match.group(1))) + \
                            stfirsthalf[rightmostdigit_match.start(2):];
                    logger.debug(f"explode:  first half {stfirsthalf}");

                # add last in matched pair to leftmost digit following match
                leftmostdigit_match = leftmostdigitre.match(stlasthalf);
                if(leftmostdigit_match):
                    stlasthalf = stlasthalf[:leftmostdigit_match.start(1)] + \
                            str(int(match.group(2)) + int(leftmostdigit_match.group(1))) + \
                            stlasthalf[leftmostdigit_match.start(2):];
                    logger.debug(f"explode:   last half {stlasthalf}");

                # replace exploded group with 0
                st = stfirsthalf + '0' + stlasthalf;
                did = True;
                break;

    # return in list form
    return did, eval(st)
    


def split(snailfish, logger):
    logger.debug(f"split: {snailfish}");
    # find any numbers 10 or greater
    if(len(snailfish) < 2):
        return False, None
    did = False;
    if(isinstance(snailfish[0], list)):
        did, snailfish[0] = split(snailfish[0], logger);
    else:
        if(snailfish[0] >= 10):
            snailfish[0] = [math.floor(snailfish[0]/2), math.ceil(snailfish[0]/2)];
            did = True
    if not did:
        if(isinstance(snailfish[1], list)):
            did, snailfish[1] = split(snailfish[1], logger);
        else:
            if(snailfish[1] >= 10):
                snailfish[1] = [math.floor(snailfish[1]/2), math.ceil(snailfish[1]/2)];
                did = True
    
    return did, snailfish;


def reduce(snailfish, logger):
    did = True;
    while(did):
        did, snailfish = explode(snailfish, logger);
        if not did:
            did, snailfish = split(snailfish, logger);
            logger.debug(f" split result: {snailfish}");
        print(f"step: {snailfish}");
    return snailfish

def magnitude(snailfish):
    lhs_mag = snailfish[0];
    if(isinstance(snailfish[0], list)):
        lhs_mag = magnitude(snailfish[0]);
    rhs_mag = snailfish[1];
    if(isinstance(snailfish[1], list)):
        rhs_mag = magnitude(snailfish[1]);
    return 3*lhs_mag + 2*rhs_mag;


#def test(logger):
#    #input = [[[[[6,6],[6,6]],[[6,0],[6,7]]],[[[7,7],[8,9]],[8,[8,1]]]], [2, 9]]
#    #output = reduce(input, logger = logger);
#    input= [[[[12, 12], [6, 14]], [[15, 0], [17, [8, 1]]]], [2, 9]];
#    print(f"input:  {input}");
#    did,output=explode(input, logger = logger)
#    print(f"result: {output}");

def test(logger):
    nerrs = 0;

    explode_testcases = [
        ([[[[[3,4],1]]]], True, [[[[0,5]]]]),
        ([[[[[9,8],1],2],3],4], True, [[[[0,9],2],3],4]),
        ([7,[6,[5,[4,[3,2]]]]], True, [7,[6,[5,[7,0]]]]),
        ([[6,[5,[4,[3,2]]]],1], True, [[6,[5,[7,0]]],3]),
    ]

    for input, did, output in explode_testcases:
        calc_did, calc_output = explode(input, logger = logger);
        if(calc_did != did):
            print(f'testcase {str(input)}, expected explode {did}, got {calc_did}');
            nerrs += 1;
        elif(calc_output != output):
            print(f'testcase {str(input)}, expected output {output}, got {calc_output}');
            nerrs += 1;

    reduce_testcases = [
            ([[[[1,1],[2,2]],[3,3]],[4,4]], [[[[1,1],[2,2]],[3,3]],[4,4]]),
            ([[[[[1,1],[2,2]],[3,3]],[4,4]],[5,5]], [[[[3,0],[5,3]],[4,4]],[5,5]]),
            ([[[[[[1,1],[2,2]],[3,3]],[4,4]],[5,5]],[6,6]], [[[[5,0],[7,4]],[5,5]],[6,6]]),
            ([[[[0,[4,5]],[0,0]],[[[4,5],[2,6]],[9,5]]],[7,[[[3,7],[4,3]],[[6,3],[8,8]]]]], [[[[4,0],[5,4]],[[7,7],[6,0]]],[[8,[7,7]],[[7,9],[5,0]]]]),
            ([[[[[4,0],[5,4]],[[7,7],[6,0]]],[[8,[7,7]],[[7,9],[5,0]]]]
                ,[[2,[[0,8],[3,4]]],[[[6,7],1],[7,[1,6]]]]], [[[[6,7],[6,7]],[[7,7],[0,7]]],[[[8,7],[7,7]],[[8,8],[8,0]]]]),
            ([[[[[6,7],[6,7]],[[7,7],[0,7]]],[[[8,7],[7,7]],[[8,8],[8,0]]]],[[[[2,4],7],[6,[0,5]]],[[[6,8],[2,8]],[[2,1],[4,5]]]]],
                [[[[7,0],[7,7]],[[7,7],[7,8]]],[[[7,7],[8,8]],[[7,7],[8,7]]]]),
            ([[[[[7,0],[7,7]],[[7,7],[7,8]]],[[[7,7],[8,8]],[[7,7],[8,7]]]],[7,[5,[[3,8],[1,4]]]]],
                [[[[7,7],[7,8]],[[9,5],[8,7]]],[[[6,8],[0,8]],[[9,9],[9,0]]]]),

            ([[[[[7,7],[7,8]],[[9,5],[8,7]]],[[[6,8],[0,8]],[[9,9],[9,0]]]],[[2,[2,2]],[8,[8,1]]]],
                [[[[6,6],[6,6]],[[6,0],[6,7]]],[[[7,7],[8,9]],[8,[8,1]]]]),

            ([[[[[6,6],[6,6]],[[6,0],[6,7]]],[[[7,7],[8,9]],[8,[8,1]]]],[2,9]],
                [[[[6,6],[7,7]],[[0,7],[7,7]]],[[[5,5],[5,6]],9]]),

            #([[[[[6,6],[7,7]],[[0,7],[7,7]]],[[[5,5],[5,6]],9]],[1,[[[9,3],9],[[9,0],[0,7]]]]],
            #    [[[[7,8],[6,7]],[[6,8],[0,8]]],[[[7,7],[5,0]],[[5,5],[5,6]]]]),

            #([[[[[7,8],[6,7]],[[6,8],[0,8]]],[[[7,7],[5,0]],[[5,5],[5,6]]]],[[[5,[7,4]],7],1]],
            #    [[[[7,7],[7,7]],[[8,7],[8,7]]],[[[7,0],[7,7]],9]]),

            #([[[[[7,7],[7,7]],[[8,7],[8,7]]],[[[7,0],[7,7]],9]],[[[[4,2],2],6],[8,7]]],
            #    [[[[8,7],[7,7]],[[8,6],[7,7]]],[[[0,7],[6,6]],[8,7]]]),
        ]

    '''
[[[[0,[4,5]],[0,0]],[[[4,5],[2,6]],[9,5]]],[7,[[[3,7],[4,3]],[[6,3],[8,8]]]]]
[[[[4,0],[5,0]],[[[4,5],[2,6]],[9,5]]],[7,[[[3,7],[4,3]],[[6,3],[8,8]]]]]
[[[[4,0],[5,4]],[[0,[7,6]],[9,5]]],[7,[[[3,7],[4,3]],[[6,3],[8,8]]]]]
[[[[4,0],[5,4]],[[7,0],[15,5]]],[7,[[[3,7],[4,3]],[[6,3],[8,8]]]]]
[[[[4,0],[5,4]],[[7,15],0]],[12,[[[3,7],[4,3]],[[6,3],[8,8]]]]]
[[[[4,0],[5,4]],[[7,15],0]],[15,[[0,[11,3]],[[6,3],[8,8]]]]]
[[[[4,0],[5,4]],[[7,15],0]],[15,[[11,0],[[9,3],[8,8]]]]]
[[[[4,0],[5,4]],[[7,15],0]],[15,[[11,9],[0,[11,8]]]]]
[[[[4,0],[5,4]],[[7,15],0]],[15,[[11,9],[11,0]]]]
[[[[4,0],[5,4]],[[7,[7, 8]],0]],[15,[[11,9],[11,0]]]]
[[[[4,0],[5,4]],[[14,0],8]],[15,[[11,9],[11,0]]]]
[[[[4,0],[5,4]],[[[7, 7],0],8]],[15,[[11,9],[11,0]]]]
[[[[4,0],[5,11]],[[0,7],8]],[15,[[11,9],[11,0]]]]
[[[[4,0],[5,11]],[[0,7],8]],[[7, 8],[[11,9],[11,0]]]]
[[[[4,0],[5,11]],[[0,7],8]],[[7, 8],[[[5, 6],9],[11,0]]]]
[[[[4,0],[5,11]],[[0,7],8]],[[7, 13],[[0,15],[11,0]]]]
[[[[4,0],[5,11]],[[0,7],8]],[[7, [6, 7]],[[0,15],[11,0]]]]
[[[[4,0],[5,11]],[[0,7],8]],[[13, 0],[[7,15],[11,0]]]]
[[[[4,0],[5,11]],[[0,7],8]],[[[6,7], 0],[[7,15],[11,0]]]]
[[[[4,0],[5,11]],[[0,7],8]],[[[6,7], 0],[[7,[7,8]],[11,0]]]]
[[[[4,0],[5,11]],[[0,7],8]],[[[6,7], 0],[[14,0],[19,0]]]]
[[[[4,0],[5,11]],[[0,7],8]],[[[6,7], 0],[[[7,7],0],[19,0]]]]
[[[[4,0],[5,11]],[[0,7],8]],[[[6,7], 7],[[0,7],[19,0]]]]
[[[[4,0],[5,11]],[[0,7],8]],[[[6,7], 7],[[0,7],[[9,10],0]]]]
[[[[4,0],[5,11]],[[0,7],8]],[[[6,7], 7],[[0,16],[0,10]]]]
[[[[4,0],[5,11]],[[0,7],8]],[[[6,7], 7],[[0,[8,8]],[0,10]]]]
[[[[4,0],[5,11]],[[0,7],8]],[[[6,7], 7],[[8,8],[8,10]]]]
[[[[4,0],[5,11]],[[0,7],8]],[[[6,7], 7],[[8,8],[8,[5,5]]]]]
[[[[4,0],[5,11]],[[0,7],8]],[[[6,7], 7],[[8,8],[13,0]]]]
[[[[4,0],[5,11]],[[0,7],8]],[[[6,7], 7],[[8,8],[[6,7],0]]]]
[[[[4,0],[5,11]],[[0,7],8]],[[[6,7], 7],[[8,14],[0,7]]]]
[[[[4,0],[5,11]],[[0,7],8]],[[[6,7], 7],[[8,[7,7]],[0,7]]]]
[[[[4,0],[5,11]],[[0,7],8]],[[[6,7], 7],[[15,0],[7,7]]]]
[[[[4,0],[5,[5,6]]],[[0,7],8]],[[[6,7], 7],[[15,0],[7,7]]]]
[[[[4,0],[10,0]],[[6,7],8]],[[[6,7], 7],[[15,0],[7,7]]]]
[[[[4,0],[[5,5],0]],[[6,7],8]],[[[6,7], 7],[[15,0],[7,7]]]]
[[[[4,5],[0,5]],[[6,7],8]],[[[6,7], 7],[[15,0],[7,7]]]]



    '''

    for input, output in reduce_testcases:
        calc_output = reduce(input, logger);
        if(calc_output != output):
            print(f'reduce testcase {str(input)}, expected output:\n   {output}\n got:\n   {calc_output}');
            nerrs += 1;

    magnitude_testcases = [
            ([[1,2],[[3,4],5]], 143),
            ([[[[0,7],4],[[7,8],[6,0]]],[8,1]], 1384),
            ([[[[1,1],[2,2]],[3,3]],[4,4]], 445),
            ([[[[3,0],[5,3]],[4,4]],[5,5]], 791),
            ([[[[5,0],[7,4]],[5,5]],[6,6]], 1137),
            ([[[[8,7],[7,7]],[[8,6],[7,7]]],[[[0,7],[6,6]],[8,7]]], 3488),
    ]

    for input, answer in magnitude_testcases:
        calc_mag = magnitude(input);
        if(calc_mag != answer):
            print(f'magnitude testcase {input}: expected {answer}, got {calc_mag}');
            nerrs += 1;

    print(f"{nerrs} errors");
    return nerrs

if('__main__' == __name__):
    import argparse;

    # deal with command line arguments
    parser = argparse.ArgumentParser(
            description = 'Advent of Code {} day {} solution'.format(
                AOC_YEAR, AOC_DAY));
    logger = logging.Logger('AOC{}d{}'.format(AOC_YEAR, AOC_DAY));
    logger.addHandler(logging.StreamHandler());
    logger.setLevel(logging.INFO);

    parser.add_argument(
        '-2', '--part2', action='store_true',
        help='part 2?', default=False);
    parser.add_argument(
        '-v', '--verbose', action='store_true',
        help='be verbose', default=False);
    parser.add_argument(
        '-T', '--test', action='store_true',
        help='run unit tests', default=False);

    args = parser.parse_args();
    if(args.verbose):
        logger.setLevel(logging.DEBUG);

    if(args.test):
        sys.exit(test(logger));

    line = sys.stdin.readline();
    snailfish = eval(line);
    line = sys.stdin.readline();
    while(line):
        logger.debug("-- adding line {}".format(line));
        snailfish = reduce([snailfish, eval(line)], logger);
        line = sys.stdin.readline();

    print(f'answer: {magnitude(snailfish)}')

File: 2021_18.rb
require 'logger'

AOCDAY = 18
AOCYEAR = 2021

$logger = Logger.new(STDOUT);

class SnailfishNumber
  attr_accessor :lhs
  attr_accessor :rhs
  attr_accessor :parent
  def initialize(s, parent: nil)
    @lhs, @rhs = s
    @parent = parent

    if @lhs.class == SnailfishNumber
      @lhs.parent = self
    elsif @lhs.class == Array
      @lhs = SnailfishNumber.new(@lhs, parent: self)
    end
    if @rhs.class == SnailfishNumber
      @rhs.parent = self
    elsif @rhs.class == Array
      @rhs = SnailfishNumber.new(@rhs, parent: self)
    end
    $logger.debug("-- init #{self} parent #{parent}")
  end

  def clone()
    SnailfishNumber.new([@lhs.clone, @rhs.clone])
  end

  def explode?(depth)
    # returns true if did explode
    retval = false
    if @lhs.class == SnailfishNumber
      retval = @lhs.explode?(depth + 1)
    end
    if not retval and (@rhs.class == SnailfishNumber)
      retval = @rhs.explode?(depth + 1)
    end
    if not retval and depth >=4 and @lhs.class == Integer and @rhs.class == Integer
      # bottom-level pair; explode this
      retval = true

      # search back up for next lhs number
      searchee = self
      $logger.debug("EXPLODE start, #{searchee}");
      while not searchee.nil?
        # if we're the lhs of our parent, keep searching up
        if not searchee.parent.nil?
          $logger.debug("LHS: searchee #{searchee} parent #{searchee.parent} lhs? #{searchee.parent.lhs == searchee}")
        end
        if searchee.parent.nil? or searchee.parent.lhs == searchee
          # move up
          searchee = searchee.parent;
          $logger.debug("LHS: navigated up, searchee now #{searchee}")
        else
          # move down
          # else we're the rhs of our parent. Take the lhs; search to the
          # rightmost element; add to that number.
          searchee = searchee.parent
          if searchee.lhs.class == Integer
            $logger.debug("LHS: found next lhs, searchee #{searchee}")
            searchee.lhs += @lhs
            searchee = nil
            $logger.debug("LHS: found next lhs, altered to #{searchee}");
          else
            $logger.debug("LHS: we were the rhs of #{searchee}, now searching right down #{searchee.lhs}")
            searchee = searchee.lhs
            while searchee.rhs.class != Integer
              searchee = searchee.rhs
              $logger.debug("LHS:  searching right down #{searchee}")
            end
            searchee.rhs += @lhs
            $logger.debug("LHS: found next rhs, altered to #{searchee}");
            searchee = nil
          end
        end
      end

      # search back up for next rhs number
      searchee = self
      $logger.debug("EXPLODE RHS start, #{searchee}");
      while not searchee.nil?
        # if we're the rhs of our parent, keep searching up
        if not searchee.parent.nil?
          $logger.debug("RHS: searchee #{searchee} parent #{searchee.parent} rhs? #{searchee.parent.rhs == searchee}")
        end
        if searchee.parent.nil? or searchee.parent.rhs == searchee
          # move up
          searchee = searchee.parent;
          $logger.debug("RHS: navigated up, searchee now #{searchee}")
        else
          # move down
          # else we're the lhs of our parent. Take the rhs; search to the
          # leftmost element; add to that number.
          searchee = searchee.parent
          if searchee.rhs.class == Integer
            $logger.debug("RHS: found next rhs, searchee #{searchee}")
            searchee.rhs += @rhs
            searchee = nil
            $logger.debug("RHS: found next rhs, altered to #{searchee}");
          else
            $logger.debug("RHS: we were the lhs of #{searchee}, now searching left down #{searchee.rhs}")
            searchee = searchee.rhs
            while searchee.lhs.class != Integer
              searchee = searchee.lhs
              $logger.debug("RHS:  searching left down #{searchee}")
            end
            searchee.lhs += @rhs
            $logger.debug("RHS: found next lhs, altered to #{searchee}");
            searchee = nil
          end
        end
      end

      # replace this instance with 0 in the parent
      if (@parent.rhs == self)
        @parent.rhs = 0
      else
        @parent.lhs = 0
      end
    end
    return retval
  end

  def split?()
    # check for splits: perform the first split
    # return true if split
    retval = false
    if @lhs.class == Integer
      if @lhs >= 10
        # do split
        @lhs = SnailfishNumber.new([@lhs.div(2), (@lhs + 1).div(2)], parent: self)
        retval = true
      end
    else
      # lhs is not an integer; must be a SnailfishNumber
      retval = @lhs.split?
    end

    if not retval
      # same thing on the rhs
      if @rhs.class == Integer
        if @rhs >= 10
          @rhs = SnailfishNumber.new([@rhs.div(2), (@rhs + 1).div(2)], parent: self)
          retval = true
        end
      else
        # rhs is a SnailfishNumber
        retval = @rhs.split?
      end
    end

    return retval
  end

  def snail_reduce!()
    $logger.debug("reducing #{self}")
    # reduce this number
    didsomething = true
    while didsomething
      didsomething = explode?(0)
      if not didsomething
        didsomething = split?
        if didsomething
          $logger.debug(" after split:   #{self}")
        end
      else
        $logger.debug(" after explode: #{self}")
      end
    end
  end

  def add(n)
    # add in a new snailfish number
    # returns a new snailfish number
    x = SnailfishNumber.new([self, n])
    x.snail_reduce!
    return x
  end

  def to_s()
    return "[#{@lhs.to_s},#{@rhs.to_s}]"
  end

  def magnitude()
    lvalue = @lhs
    if SnailfishNumber == @lhs.class
      lvalue = @lhs.magnitude
    end
    rvalue = @rhs
    if SnailfishNumber == @rhs.class
      rvalue = @rhs.magnitude
    end

    #$logger.debug("----- #{3*lvalue + 2*rvalue} from l #{lvalue} r #{rvalue} from #{self}")
    3*lvalue + 2*rvalue
  end
end

#x = SnailfishNumber.new([[2,[[7,7],7]],[[5,8],[[9,3],[0,2]]]])
#y = SnailfishNumber.new([[[0,[5,8]],[[1,7],[9,6]]],[[4,[1,2]],[[1,4],2]]])
#puts("adding #{x} to #{y}")
#z = x.add(y)
#puts("  #{z}")
#puts("Mag #{z.magnitude}")
#
#z = SnailfishNumber.new([[[[7,8],[6,6]],[[6,0],[7,7]]],[[[7,8],[8,8]],[[7,9],[0,6]]]])
#puts("  #{z}")
#puts("Mag #{z.magnitude}")
#raise NoSuchError


def run(inputfile, **kwargs)
  numbers = []
  File.open(inputfile).each_line do |line|
    numbers << SnailfishNumber.new(eval(line.strip))
  end

  max_magnitude = 0
  if kwargs.include?(:part2)
    # find the larges magnitude of the sum of any two numbers
    numbers.permutation(2).each do |x,y|
      xp = x.clone
      yp = y.clone
      zp = xp.add(yp)
      new_magnitude = zp.magnitude
      max_magnitude = [max_magnitude, new_magnitude].max
    end
    puts("Max pair magnitude #{max_magnitude}")
  else
    n = numbers.reduce(:add)
    puts("Final snailfish number: #{n}")
    puts("Final magnitude: #{n.magnitude}")
  end
end

if __FILE__ == $0
  require 'optparse'

  options = {}
  OptionParser.new do |opts|
    opts.banner = "Usage: $0 [options] <inputfile> [<more inputfiles>]\n" \
      "  Advent of Code #{AOCYEAR} day #{AOCDAY}"

    opts.on("-v", "--verbose", "Be verbose") do |v|
      options[:verbose] = v
    end

    opts.on("-2", "--part2", "Run part 2") do |part2|
      options[:part2] = true
    end

    opts.on("-h", "--help", "Print this help message") do
      puts opts
      exit
    end
  end.parse!

  $logger.level = (options.include? :verbose) ? Logger::DEBUG : Logger::WARN

  ARGV.each do |inputfile|
    run(inputfile, **options)
  end
end
